From 31f4f2f435b72c46c34d8970eb09e5c3a0e93254 Mon Sep 17 00:00:00 2001
From: Edmunt Pienkowsky <roed@onet.eu>
Date: Tue, 8 Jun 2021 13:26:56 +0200
Subject: [PATCH 08/13] Use gcc visibility attributes

Ability to compile sources with -fvisibility=hidden and -fvisibility-inlines-hidden flags.

diff --git a/builds/install/misc/databases.conf b/builds/install/misc/databases.conf
index cc5b62de8a..86c865eb0d 100644
--- a/builds/install/misc/databases.conf
+++ b/builds/install/misc/databases.conf
@@ -15,10 +15,14 @@
 #
 
 #
-# Example Database:
+# Your first database:
 #
-employee.fdb = $(dir_sampleDb)/employee.fdb
-employee = $(dir_sampleDb)/employee.fdb
+my-first = /var/lib/firebird/data/my-first.fdb
+{
+	RemoteAccess = false
+	DefaultDbCachePages = 50
+	SecurityDatabase = my-first
+}
 
 #
 # Master security database specific setup.
diff --git a/src/common/common.h b/src/common/common.h
index 9fbe95410a..efd87cd4c2 100644
--- a/src/common/common.h
+++ b/src/common/common.h
@@ -207,6 +207,13 @@
 #define FB_CPU CpuM68k
 #endif /* M68K */
 
+#ifdef __GNUC__
+#define API_ROUTINE __attribute__((visibility("default")))
+#define API_ROUTINE_VARARG API_ROUTINE
+#define INTERNAL_API_ROUTINE API_ROUTINE
+#define FB_EXPORTED __attribute__((visibility("default")))
+#endif
+
 #endif /* LINUX */
 
 
diff --git a/src/include/firebird.h b/src/include/firebird.h
index 1bd7f2d4b3..99845779c3 100644
--- a/src/include/firebird.h
+++ b/src/include/firebird.h
@@ -47,6 +47,8 @@
 #define FB_DLL_EXPORT __declspec(dllexport)
 #elif defined(DARWIN)
 #define FB_DLL_EXPORT API_ROUTINE
+#elif defined(LINUX) && defined(__GNUC__)
+#define FB_DLL_EXPORT API_ROUTINE
 #else
 #define FB_DLL_EXPORT
 #endif
diff --git a/src/intl/ld.cpp b/src/intl/ld.cpp
index 1056921ee3..c22081beed 100644
--- a/src/intl/ld.cpp
+++ b/src/intl/ld.cpp
@@ -468,7 +468,7 @@ struct
 };
 
 
-INTL_BOOL FB_DLL_EXPORT LD_lookup_charset(charset* cs, const ASCII* name, const ASCII* /*config_info*/)
+extern "C" INTL_BOOL FB_DLL_EXPORT LD_lookup_charset(charset* cs, const ASCII* name, const ASCII* /*config_info*/)
 {
 	// ASF: We can't read config_info if version < INTL_VERSION_2,
 	// since it wasn't pushed in the stack by the engine.
@@ -491,7 +491,7 @@ INTL_BOOL FB_DLL_EXPORT LD_lookup_charset(charset* cs, const ASCII* name, const
 }
 
 
-INTL_BOOL FB_DLL_EXPORT LD_lookup_texttype(texttype* tt, const ASCII* texttype_name, const ASCII* charset_name,
+extern "C" INTL_BOOL FB_DLL_EXPORT LD_lookup_texttype(texttype* tt, const ASCII* texttype_name, const ASCII* charset_name,
 										   USHORT attributes, const UCHAR* specific_attributes,
 										   ULONG specific_attributes_length, INTL_BOOL ignore_attributes,
 										   const ASCII* config_info)
@@ -557,7 +557,7 @@ INTL_BOOL FB_DLL_EXPORT LD_lookup_texttype(texttype* tt, const ASCII* texttype_n
 }
 
 
-ULONG FB_DLL_EXPORT LD_setup_attributes(
+extern "C" ULONG FB_DLL_EXPORT LD_setup_attributes(
 	const ASCII* textTypeName, const ASCII* charSetName, const ASCII* configInfo,
 	ULONG srcLen, const UCHAR* src, ULONG dstLen, UCHAR* dst)
 {
@@ -583,7 +583,7 @@ ULONG FB_DLL_EXPORT LD_setup_attributes(
 }
 
 
-void FB_DLL_EXPORT LD_version(USHORT* version)
+extern "C" void FB_DLL_EXPORT LD_version(USHORT* version)
 {
 	// We support version 1 and 2.
 	if (*version != INTL_VERSION_1)
diff --git a/src/intl/ld.h b/src/intl/ld.h
index 0ee001f733..8020254d2d 100644
--- a/src/intl/ld.h
+++ b/src/intl/ld.h
@@ -55,15 +55,6 @@
 
 #define UINT16	USHORT
 
-#if defined(WIN_NT)
-#define FB_DLL_EXPORT __declspec(dllexport)
-#elif defined(DARWIN)
-#define FB_DLL_EXPORT API_ROUTINE
-#else
-#define FB_DLL_EXPORT
-#endif
-
-
 /* Following this line is LD.H from Borland Language Driver Kit */
 
 
